src/mococrw/kdf.h

```diff
@@ -0,0 +1,117 @@
+     */
+    virtual ~KeyDerivationFunction();
+    /**
+     * @brief Provides a derived key.
+     *
+     * This function returns a derived key based on the arguments to this function and the arguments provided to
+     * the constructor
+     * @param password The password/secret key which shall be used for key derivation
+     * @param outputLength The length of the derived key in bytes.
+     * @param salt The salt used for the key derivation
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error.
+     * @return The derived key
+     */
+    virtual std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                           const std::vector<uint8_t> &salt) = 0;
+};
+
+class PBKDF2: public KeyDerivationFunction {
+public:
+    /**
+     * @brief Constructor
+     * @param hashFunction The hash function which shall be used
+     */
+    PBKDF2(openssl::DigestTypes hashFunction, uint32_t iterations);
+
+    ~PBKDF2();
+
+    /**
+     * @brief Provides a derived key.
+     *
+     * This function returns a derived key based on the arguments to this function and the arguments provided to
+     * the constructor
+     * @param password The password/secret key which shall be used for key derivation
+     * @param outputLength The length of the derived key in bytes.
+     * @param salt The salt used for the key derivation
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error.
+     * @return The derived key
+     */
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                   const std::vector<uint8_t> &salt) override;
+
+private:
+    openssl::DigestTypes hashFunction;
+    uint32_t iterations;
+    std::vector<uint8_t>_derivedKey;
+
+};
+
+class X963KDF : public KeyDerivationFunction
+{
+public:
+    /**
+     * @brief Constructor
+     * @param hashFunction The hash function which shall be used
+     */
+    X963KDF(openssl::DigestTypes hashFunction);
+
+    /**
+     * @brief destructor
+     */
+    ~X963KDF();
+
+    /**
+     * @brief Provides a derived key.
+     *
+     * This function returns a derived key based on the arguments to this function and the arguments provided to
+     * the constructor
+     * @param password The password/secret key which shall be used for key derivation
+     * @param outputLength The length of the derived key in bytes.
+     * @param salt The salt used for the key derivation
+     * @throws This method may throw an OpenSSLException if OpenSSL indicates an error.
+     * @return The derived key
+     */
+    std::vector<uint8_t> deriveKey(const std::vector<uint8_t> &password, const size_t outputLength,
+                                   const std::vector<uint8_t> &salt) override;
+
+private:
+    openssl::DigestTypes hashFunction;
+    uint32_t iterations;
+    std::vector<uint8_t>_derivedKey;
```

Comment:
```suggestion
    std::vector<uint8_t> _derivedKey;
```

Ditto, do we really need this?

Reply:
done